test/proxy_test.go

```diff
@@ -0,0 +1,606 @@
+		return fmt.Errorf("Proxy server received Connect req %v, but should not", req)
+	})
+
+	// Create a custom dialer that directly dials the backend.  We'll use this
+	// to bypass any proxy logic.
+	dialerCalled := make(chan bool, 1)
+	customDialer := func(_ context.Context, backendAddr string) (net.Conn, error) {
+		dialerCalled <- true
+		return net.Dial("tcp", backendAddr)
+	}
+
+	targetResolver := setupDNS(t)
+	targetResolver.InitialState(resolver.State{Addresses: []resolver.Address{{Addr: backendAddr}}})
+
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithContextDialer(customDialer),
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	conn, err := grpc.NewClient(unresolvedTargetURI, dopts...)
+	if err != nil {
+		t.Fatalf("grpc.NewClient() failed: %v", err)
+	}
+	defer conn.Close()
+
+	client := testgrpc.NewTestServiceClient(conn)
+	if _, err := client.EmptyCall(ctx, &testgrpc.Empty{}); err != nil {
+		t.Errorf("EmptyCall() failed: %v", err)
+	}
+
+	select {
+	case <-dialerCalled:
+		t.Logf("custom dialer was called by grpc.NewClient()")
+	default:
+		t.Errorf("custom dialer was not called by grpc.NewClient()")
+	}
+
+	select {
+	case <-proxyStartedCh:
+		t.Fatalf("Proxy Server was dialled")
+	default:
+		t.Logf("Proxy Server was not dialled")
+	}
+}
+
+// Tests grpc.NewClient with default i.e DNS resolver for targetURI and a proxy
+// and verifies that it connects to proxy server and sends unresolved target
+// URI in the HTTP CONNECT req and connects to backend. Also verifies that
+// correct user info is sent in the CONNECT.
+func (s) TestBasicAuthInGrpcNewClientWithProxy(t *testing.T) {
+	// Set up a channel to receive signals from OnClientResolution.
+	resolutionCh := make(chan bool, 1)
+	// Overwrite OnClientResolution to send a signal to the channel.
+	origOnClientResolution := delegatingresolver.OnClientResolution
+	delegatingresolver.OnClientResolution = func(int) {
+		resolutionCh <- true
+	}
+	t.Cleanup(func() { delegatingresolver.OnClientResolution = origOnClientResolution })
+
+	// Create and start a backend server.
+	backendAddr := createAndStartBackendServer(t)
+	const (
+		user     = "notAUser"
+		password = "notAPassword"
+	)
+	// Set up and start the proxy server.
+	proxyLis, errCh, doneCh, _ := setupProxy(t, backendAddr, false, func(req *http.Request) error {
+		if req.Method != http.MethodConnect {
+			return fmt.Errorf("unexpected Method %q, want %q", req.Method, http.MethodConnect)
+		}
+		if req.URL.Host != unresolvedTargetURI {
+			return fmt.Errorf("unexpected URL.Host %q, want %q", req.URL.Host, unresolvedTargetURI)
+		}
+		wantProxyAuthStr := "Basic " + base64.StdEncoding.EncodeToString([]byte(user+":"+password))
+		if got := req.Header.Get("Proxy-Authorization"); got != wantProxyAuthStr {
+			gotDecoded, _ := base64.StdEncoding.DecodeString(got)
```

Comment:
we should not ignore decoding errors. it should return error if decoding fails

Reply:
Done.